### ISSUE TITLE
FR: Flatpak/Flathub Support (Bounty Included)

### ISSUE THREAD

**Issue by anthonyajumelles** [koreader/koreader#6349]
As platforms like the PinePhone and Fedora Silverblue begin to take more prominence alongside the widespread adoption of Flatpak for all major distributions, including ChromeOS, relying on solely AppImage and Debs increases the friction for end-users.

Automatic updates as well as ensuring all needed libraries are present are two major benefits that KOReader could observe.

In that spirit, I have begun a bounty to implement Flatpak!

https://www.bountysource.com/issues/91941744-fr-flatpak-flathub-support-bounty-included

**Comment by Frenzie** [comment]
As an aside, your link says:

{"error":"Internal server error."}

It probably shouldn't be too hard, but back when I added the AppImage that was the one with clear documentation and the other two were basically severely incomplete technogibberish. ;-)

**Comment by pazos** [comment]
> including ChromeOS

They way to install linux apps on ChromeOS is still tricky: https://flatpak.org/setup/Chrome%20OS/ and instead of using `sudo apt install flatpak`  and add flatpak repos one could use `sudo dpkg -i koreader.deb`

But, even better, the android APK will be compatible with ChromeOS soon. That reduces the friction as you said.

Here's a screenshot: 
![2020-07-05_14-24](https://user-images.githubusercontent.com/975883/86535878-af87fd00-bee3-11ea-898f-fac0ef3f25f3.png)

The only downside of this is the lack of devices to test. The emulator is still a beta and is *very* broken. So it is not possible to do [recommended test-cases](https://developer.android.com/topic/arc/tests). If somebody with a real Chromebook  want to test then please be my guess :smile:

**Comment by anthonyajumelles** [comment]
@pazos the APK still requires entering developer mode on Chromebooks no? Or is KOReader going to be on the Play Store?

**Comment by NiLuJe** [comment]
@poire-z: I do have a recent-enough Chromebook to have Linux support, yeah.

**Comment by pazos** [comment]
> the APK still requires entering developer mode on Chromebooks no?

Never played with a Chromebook. On the emulator is just a matter of dragging the APK on top of the emulator window.

> Or is KOReader going to be on the Play Store?

I don't think so. We don't have 64 bit support (x86_64 and aarch64) that are required to publish on the Play Store. Other than that we -magically- don't trigger any other warning on play developer console. So it should be possible.

>  I do have a recent-enough Chromebook to have Linux support

The man with a million devices :p

**Comment by Frenzie** [comment]
@pazos While I have no particular inclination to, couldn't I just run that ChromeOS thing on any of my laptops?

**Comment by pazos** [comment]
Yeah, with something like https://anbox.io/, which implement android runtime in containers, just like ChromeOS. But that doesn't work for some test cases.

**Comment by Frenzie** [comment]
That's not quite what I meant (although for our purposes it's more or less the same), but I guess I was wrong regardless.

https://www.androidcentral.com/how-convert-windows-laptop-chromebook
> You can't just download Chrome OS and install it on any laptop like you can Windows and Linux. Chrome OS is closed source and only available on proper Chromebooks. But Chromium OS is 90% the same as Chrome OS. More importantly, it's open source: you can download Chromium OS and build on top of it if you so choose.

**Comment by NiLuJe** [comment]
I have no idea how feasible that is on random hardware, but Chrome OS is essentially Chromium+Google installed in a custom Gentoo-based Linux distro ;).

**Comment by connlim** [comment]
Hi, I've made a flatpak manifest for KOReader. It uses the latest .deb release to generate the flatpak build. I tried making the flatpak with the appimage instead but kept running into crashes. 

Source code: https://github.com/connlim/flathub/tree/rocks.koreader.KOReader

This is my first time deploying to flatpak so there may be mistakes. If anybody here is more experienced with flatpak you are also welcome to take my source code and build on it. Take a look and tell me if there's any issues. Thanks!

It should already be deployable to flathub (I hope). I didn't do it myself as the developer should be the one making the PR to flathub.

To test locally:
```bash
# Export the flatpak into a local repo
flatpak-builder --repo=my-repo build-dir rocks.koreader.KOReader.yaml

# Install the flatpak from the repo
flatpak install ./my-repo rocks.koreader.KOReader
```
References:
- https://docs.flatpak.org/en/latest/first-build.html#test-the-build
- https://docs.flatpak.org/en/latest/flatpak-builder.html

**Comment by Frenzie** [comment]
@connlim That's an interesting approach, and it does essentially come down to the same thing. I was thinking more in the direction of adding an extra target in the Makefile, but if simply repackaging works well enough then I suppose I don't see why not.

https://github.com/koreader/koreader/blob/8198b2bcc3cfe8ace343fd32eddc673b33ab702b/Makefile#L314-L349

https://github.com/koreader/koreader/blob/8198b2bcc3cfe8ace343fd32eddc673b33ab702b/Makefile#L377-L401

**Comment by connlim** [comment]
Yes it should be possible to add a flatpak build target to the makefile. The issue comes when figuring out how to run the makefile and distribute the result. You have 2 main options:

## 1. Tell Flathub to clone the repo and run the build target
The issue with this method is that [Flathub does not allow network access during the build process](https://github.com/flathub/flathub/wiki/App-Requirements#stable-releases-reproducible-builds). All files have to be downloaded under the `sources` section before the build process can start.

From my attempts at doing this it seems that [koreader-base relies on internet access to download third party packages during the build process](https://github.com/koreader/koreader-base/tree/master/thirdparty). Refactoring all that code to download the third party packages via git submodules (allowing flatpak to download them automatically before building) would be too time consuming.

## 2. Generate a .flatpak file with each github release
This can be done by creating a [single-file bundle](https://docs.flatpak.org/en/latest/single-file-bundles.html). However you can't simply upload this .flatpak to Flathub. Flathub does not distribute .flatpak files directly. Instead, it runs a bot which builds the package according to a manifest. **So we would still need to write a manifest for the Flathub bot to download and extract the .flatpak file**.

The only benefit of generating a .flatpak file is that it would also allow users to install the package without access to Flathub.

## Personal opinion (take with a pinch of salt)
In my opinion repackaging the .deb is still the best method. This is actually quite common: many packages on flathub do the same thing. Building a .flatpak file is useless duplicate work when simply extracting files from the .deb works fine. 

Furthermore, having a .flatpak file for distribution is pointless when you already have AppImages. The main advantage of flatpak **IS** Flathub. Most users get flatpak packages from Flathub. Even if a user downloads the .flatpak file, they are still required to have flatpak and the freedesktop runtimes (downloaded from Flathub) installed before it can work. For portable distribution AppImage is still superior. *(Note: technically [it is possible to include the runtimes](https://docs.flatpak.org/en/latest/usb-drives.html), but the file would be obscenely big)* 
  
Unfortunately I am not well-versed with makefiles so I won't be able to help you add flatpak as a build target *(I tried and failed miserably)*. If anybody wishes to try, feel free to take [my work](https://github.com/connlim/flathub/tree/rocks.koreader.KOReader) and build on it.

**Comment by pazos** [comment]
> In my opinion repackaging the .deb is still the best method

I agree. We could provide a standard tar.gz with the same binaries to avoid some steps of the manifest and the patch, but I don't know if that's a bit overkill.

> Even if a user downloads the .flatpak file, they are still required to have flatpak and the freedesktop runtimes

I'm not familiar with flathub or their runtimes. Is the freedesktop runtime needed only by SDL2?.

**Comment by Frenzie** [comment]
> In my opinion repackaging the .deb is still the best method. This is actually quite common: many packages on flathub do the same thing. Building a .flatpak file is useless duplicate work when simply extracting files from the .deb works fine.

Not really any more duplicate than what you posted since it's just a few minor steps at the end that differ, regardless if you do it straight away or by modifying a DEB. But like I said, I wasn't objecting. ^_^

> Most users get flatpak packages from Flathub.

Fwiw, I have a handful installed, none of them from Flathub. (Besides some dependencies it pulls in from there.)

> Unfortunately I am not well-versed with makefiles so I won't be able to help you add flatpak as a build target (I tried and failed miserably). If anybody wishes to try, feel free to take my work and build on it. In any case, I do hope deploying koreader to flathub succeeds. It would be great for Linux phones.

Probably not without the ARM files. Is there a way to make it use all three DEB files for the different architectures?

@pazos
> I'm not familiar with flathub or their runtimes. Is the freedesktop runtime needed only by SDL2?.

See https://docs.flatpak.org/en/latest/available-runtimes.html

**Comment by connlim** [comment]
> Not really any more duplicate than what you posted since it's just a few minor steps at the end that differ, regardless if you do it straight away or by modifying a DEB. But like I said, I wasn't objecting. ^_^

Yeah you're right it's possible to avoid duplication of work since the flatpak and deb are more or less the same thing. Maybe someone could try implementing it this way?

> Probably not without the ARM files. Is there a way to make it use all three DEB files for the different architectures?

It should be possible to run it on arm. [I added an untested commit which should enable it](https://github.com/connlim/flathub/commit/8b0eabae5b300ee6374cab65339b6223219f85ba). It would be good if someone with an arm machine could try building it.

**Comment by pazos** [comment]
If the phones are from the future I guess they would be all arm64 :smile: They should work fine with armhf anyways.

**Comment by farvardin** [comment]
I hope flatpaks will never replace appimages (which is already working just fine)

**Comment by rokejulianlockhart** [comment]
@pazos, regarding https://github.com/koreader/koreader/issues/6349#issuecomment-653901775:~:text=the%20only%20downside%20of%20this%20is%20the%20lack%20of%20devices%20to%20test., have you tried the recently released ChromeOS Flex?

**Comment by in-plaintext** [comment]
there's now an official package: https://flathub.org/apps/rocks.koreader.KOReader

**Comment by Frenzie** [comment]
See https://github.com/flathub/rocks.koreader.KOReader for the source, thanks to @proletarius101
